File: DailyChallenges/Scala/LeetCode_150.scala
object Solution {
    def evalRPN(tokens: Array[String]): Int = {
        def eval(i: Int = 0, stack: List[Int] = Nil): Int = 
            if (i == tokens.length) stack.head 
            else tokens(i) match {
                case "+" => eval(i+1, (stack.tail.head + stack.head) :: stack.tail.tail)
                case "-" => eval(i+1, (stack.tail.head - stack.head) :: stack.tail.tail)
                case "*" => eval(i+1, (stack.tail.head * stack.head) :: stack.tail.tail)
                case "/" => eval(i+1, (stack.tail.head / stack.head) :: stack.tail.tail)
                case str => eval(i+1, str.toInt :: stack)
            }

        eval()
    }
}

File: DailyChallenges/CPP/LeetCode_150.cpp
class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        vector<long long> stack;
        for (auto &s: tokens){
            if (all_of(s.begin(), s.end(), ::isdigit) || (s[0] == '-' && s.size()>1))
                stack.push_back(stoi(s));
            else{
                int b = stack.back(); stack.pop_back();
                int a = stack.back(); stack.pop_back();
                if (s == "+")
                    stack.push_back(a+b);
                else if (s == "-")
                    stack.push_back(a-b);
                else if (s == "*")
                    stack.push_back((long long)a*b);
                else
                    stack.push_back(a/b);
            }
        }
        return stack.front();
    }
};
